eth/vm/forks/london/state.py

```diff
@@ -0,0 +1,193 @@
+        gas_refund_amount = (gas_refund + gas_remaining) * transaction_context.gas_price
+
+        if gas_refund_amount:
+            self.vm_state.logger.debug2(
+                'TRANSACTION REFUND: %s -> %s',
+                gas_refund_amount,
+                encode_hex(computation.msg.sender),
+            )
+
+            self.vm_state.delta_balance(computation.msg.sender, gas_refund_amount)
+
+        # Miner Fees
+        transaction_fee = \
+            (transaction.gas - gas_remaining - gas_refund) * transaction.max_priority_fee_per_gas
+        self.vm_state.logger.debug2(
+            'TRANSACTION FEE: %s -> %s',
+            transaction_fee,
+            encode_hex(self.vm_state.coinbase),
+        )
+        self.vm_state.delta_balance(self.vm_state.coinbase, transaction_fee)
+
+        # Process Self Destructs
+        for account, _ in computation.get_accounts_for_deletion():
+            # TODO: need to figure out how we prevent multiple selfdestructs from
+            # the same account and if this is the right place to put this.
+            self.vm_state.logger.debug2('DELETING ACCOUNT: %s', encode_hex(account))
+
+            # TODO: this balance setting is likely superflous and can be
+            # removed since `delete_account` does this.
+            self.vm_state.set_balance(account, 0)
+            self.vm_state.delete_account(account)
+
+        return computation
+
+class LondonState(BerlinState):
+    account_db_class: Type[AccountDatabaseAPI] = AccountDB
+    computation_class = LondonComputation
+    transaction_executor_class: Type[TransactionExecutorAPI] = LondonTransactionExecutor
+
+    def validate_transaction(
+        self,
+        transaction: SignedTransactionAPI
+    ) -> None:
+        # homestead validation
+        if transaction.s > SECPK1_N // 2 or transaction.s == 0:
+            raise ValidationError("Invalid signature S value")
+
+        normalized_transaction = normalize_transaction(transaction)
+        validate_london_normalized_transaction(
+            state=self,
+            transaction=normalized_transaction,
+            base_fee_per_gas=self.execution_context.base_gas_fee
+        )
+
+    def get_transaction_context(self: StateAPI,
+                                transaction: SignedTransactionAPI) -> TransactionContextAPI:
+        """
+        London-specific transaction context creation,
+        where gas_price includes the block base fee
+        """
+        if isinstance(transaction, LondonTypedTransaction):  # TODO probably do this somewhere else
+            priority_fee_per_gas = min(
+                transaction.max_priority_fee_per_gas,
+                transaction.max_fee_per_gas - self.execution_context.base_gas_fee
+            )
+        else:
+            priority_fee_per_gas = min(
+                transaction.gas_price,
+                transaction.gas_price - self.execution_context.base_gas_fee
+            )
```

Comment:
Yeah, my instinct is to put this in the transaction class, just have everything implement a `priority_fee_per_gas` attribute. Then we get to skip the instance check.